### PR TITLE
refactor(connectors): refactor sqla query generation

### PR DESCRIPTION
**Pull request by mayurnewase**
### SUMMARY
Breaking thunk "get_sqla_query" into pieces

### TODO

- [x] Remove Debugs
- [x] Fix bugs and tests
- [x] Move functions into utils
- [x] Lint
- [x] Write unit tests
- [ ] Break this in small PRs

### CODE REVIEW

**Comment by ktmud** [comment]
Could you hold until #10270 is merged? I'm afraid there will be a lot of conflicts.

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -871,102 +908,76 @@ def _get_sqla_row_level_filters(
...
-            and (is_sip_38 or (not is_sip_38 and not groupby))
-        ):
-            raise QueryObjectValidationError(_("Empty query?"))
-        metrics_exprs: List[ColumnElement] = []
+    def _get_metric_expressions(
+        self,
+        metrics: List[Metric],
+        columns_by_name: Dict[str, TableColumn],
+        metrics_by_name: Dict[str, SqlMetric],
+    ) -> List[Label]:
```
Comment (ktmud) [inline superset/connectors/sqla/models.py:925]: I wonder how many of these private methods can be move out of the class. We should also try not to pass too many arguments between functions.

Comment (mayurnewase) [reply superset/connectors/sqla/models.py:925]: Moved few,but these use class attributes/other class in this file in some way,which produce circular imports.

File: superset/connectors/sqla/models.py [hunk 2]
```diff
@@ -377,7 +389,9 @@ class SqlMetric(Model, BaseMetric):
 
     def get_sqla_col(self, label: Optional[str] = None) -> Column:
         label = label or self.metric_name
+        print("label ", label)
         sqla_col = literal_column(self.expression)
+        print("sqla col ", str(sqla_col))
```
Comment (villebro) [inline superset/connectors/sqla/models.py:394]: These print statements should probably be removed

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (villebro) [reply superset/connectors/sqla/models.py:925]: At some point we should probably also look at resolving some of those circular imports (let's leave those for another PR)

File: superset/connectors/sqla/models.py [hunk 3]
```diff
@@ -1147,109 +1125,361 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
...
+        metrics: List[Metric],
+        granularity: str,
+        filter_: Optional[List[Dict[str, Any]]] = None,
+        extras: Optional[Dict[str, Any]] = None,
+        columns: Optional[List[str]] = None,
+        groupby: Optional[List[str]] = None,
+        orderby: Optional[List[Tuple[ColumnElement, bool]]] = None,
+        inner_from_dttm: Optional[datetime] = None,
+        inner_to_dttm: Optional[datetime] = None,
+    ) -> Tuple[TableClause, List[str]]:
```
Comment (villebro) [inline superset/connectors/sqla/models.py:1206]: Some of these are available on `self`, e.g. `db_engine_spec` can be replaced with `self.database.db_engine_spec` hence doesn't need to be here.

File: superset/connectors/sqla/utils.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+from sqlalchemy.sql.expression import BooleanClauseList, Label
+
+from superset.constants import NULL_STRING
+from superset.exceptions import QueryObjectValidationError
+from superset.jinja_context import BaseTemplateProcessor
+from superset.utils.core import cast_to_num, FilterOperator
+
+
+def get_having_clause(
+    extra_having: Any, template_processor: BaseTemplateProcessor
```
Comment (villebro) [inline superset/connectors/sqla/utils.py:35]: I believe having is `Optional[str]`
```suggestion
    extra_having: Optional[str], template_processor: BaseTemplateProcessor
```

File: superset/connectors/sqla/utils.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+def get_expected_labels_from_select(select_expressions: List[Label]) -> List[str]:
+    return [
+        c._df_label_expected  # pylint: disable=protected-access
+        for c in select_expressions
+    ]
+
+
+def get_where_operation(  # pylint:disable=too-many-branches
+    flt: Dict[str, Any], operation: str, col_obj: Any, value: Any
```
Comment (villebro) [inline superset/connectors/sqla/utils.py:63]: I believe we can add some types here, too e.g. `col_obj` is probably `Column`.

**Review by villebro — COMMENTED** [review]
First pass comments. Great first step to break down the SQLA model! There will most likely be much refactoring on this model in the near term, so it would be best to keep the PRs as short/modular as possible, preferably one function broken out + unit tests added per PR. This way it's easier to review the changes and also ensure that conflicts are less likely and easier to resolve.

File: superset/connectors/sqla/utils.py (hunk 5 above)
Comment (mayurnewase) [reply superset/connectors/sqla/utils.py]: updated

File: superset/connectors/sqla/utils.py (hunk 4 above)
Comment (mayurnewase) [reply superset/connectors/sqla/utils.py]: updated

File: superset/connectors/sqla/models.py (hunk 3 above)
Comment (mayurnewase) [reply superset/connectors/sqla/models.py]: I can try to break this into pieces.

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (mayurnewase) [reply superset/connectors/sqla/models.py:923]: @villebro so should we wait for #10270 to merge or this could be merged?

**Review by ktmud — COMMENTED** [review]
+1 on breaking this into smaller PRs if possible...

I'm also wondering if in the future we could introduce a `QueryGenerator` class that translates a `QueryObject` to SQL queries piece by piece, with an API similar to the SQLAlchemy's own `Query` object. This should save the trouble of passing many parameters around multiple functions.

```python
QueryGenerator(dataset)
  .select(metrics=metrics, columns=columns)
  .filter_by(filters)
  .limit(row_limit)
  .offset(row_offset)
```

**Pull request closed without merging**